### PR TITLE
[perf] Canonicalize defining anchor for queries

### PR DESCRIPTION
**Pull request by compiler-errors**
Investigating perf implications of canonicalizing defining anchor properly...

r? @ghost cc @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
I guess this needs to wait until the anchor is not an item `DefId`, but a list of opaque type `DefId`

File: compiler/rustc_infer/src/infer/region_constraints/mod.rs [hunk 1]
```diff
@@ -442,6 +442,12 @@ impl<'tcx> RegionConstraintCollector<'_, 'tcx> {
         // cannot add constraints once regions are resolved
         debug!("RegionConstraintCollector: add_constraint({:?})", constraint);
 
+        if format!("{constraint:?}")
+            == "VarSubReg('?25, ReFree(DefId(0:19 ~ implied_bounds_from_types[4e59]::extend_lifetime), BrNamed(DefId(0:20 ~ implied_bounds_from_types[4e59]::extend_lifetime::'a), 'a)))"
+        {
+            bug!();
+        }
+
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/region_constraints/mod.rs]: I'm an idiot lmao.

**Comment by compiler-errors** [comment]
Huh, I wonder what perf looks like without that `format!()`...

@bors try @rust-timer queue

**Comment by compiler-errors** [comment]
I don't believe I have very much time to mess around with this PR any more, sorry 😓

**Pull request closed without merging**